subsys/bluetooth/host/buf.c

```diff
@@ -147,3 +150,87 @@ struct net_buf_pool *bt_buf_get_num_complete_pool(void)
 }
 #endif /* CONFIG_BT_CONN || CONFIG_BT_ISO */
 #endif /* ZTEST_UNITTEST */
+
+/* Create a "view" or "window" into an existing buffer.
+ * - enforces one active view at a time per-buffer
+ * -> this restriction enables prepending data (ie. for headers)
+ * - forbids appending data to the view
+ * - pulls the size of the view from said buffer.
+ *
+ * The "virtual buffer" that is generated has to be allocated from a buffer
+ * pool. This is to allow refcounting and attaching a destroy callback.
+ * The configured size of the buffers in that pool should be zero-length.
+ *
+ * The user-data size is application-dependent, but should be minimized to save
+ * memory. user_data is not used by the view API.
+ *
+ * The view mechanism needs to store extra metadata in order to unlock the
+ * original buffer when the view is destroyed.
+ *
+ * The storage and allocation of the view buf pool and the view metadata is the
+ * application's responsibility.
+ */
+struct net_buf *bt_buf_make_view(struct net_buf *view, struct net_buf *parent,
+				 size_t winsize, size_t headroom, struct bt_buf_view_meta *meta)
+{
+	__ASSERT_NO_MSG(winsize);
+	__ASSERT_NO_MSG(view);
+	/* The whole point of this API is to allow prepending data. If the
+	 * headroom is 0, that will not happen.
+	 */
+	__ASSERT_NO_MSG(net_buf_headroom(parent) > 0);
+
+	/* check that we actually have enough headroom */
+	__ASSERT_NO_MSG(headroom <= net_buf_headroom(parent));
+
+	/* `parent` should have been just re-used instead of trying to make a
+	 * view into it.
+	 */
+	__ASSERT_NO_MSG(winsize < parent->len);
+
+	__ASSERT_NO_MSG(parent->data != 0);
+	__ASSERT_NO_MSG(parent->size != 0);
+
+	LOG_DBG("make-view %p viewsize %u meta %p", view, winsize, meta);
+
+	/* The ref to `parent` is moved in by passing `parent` as argument. */
+	net_buf_simple_clone(&parent->b, &view->b);
+	view->size = winsize + parent->len + net_buf_headroom(parent);
+	view->len = winsize;
+	view->flags = NET_BUF_EXTERNAL_DATA;
+
+	/* we have a view, eat `winsize`'s worth of data from the parent */
+	(void)net_buf_pull(parent, winsize);
+
+	/* save backup & "clip" the buffer so the next `make_view` will fail */
+	meta->parent = parent;
```

Comment:
This should still be `= net_buf_ref(parent);`, I think (and corresponding `net_buf_unref()` in `destroy_view()`

Reply:
- Most of our APIs move the source net_buf reference on success. We should follow this pattern here as well.
- Functions should do the minimum work to be optimally composable. It allows the call site decide if it wants to clone or more the ref. If we force a clone in `bt_buf_make_view`, we will get an awkward `net_buf_unref` at the use site.

Sketch of awkward `net_buf_unref`:
```c
bt_buf_make_view(view, target_buf, view_size, view_headroom, view_meta);
// This is the last use of `target_buf`. But we still have the reference.
// We move it into `net_buf_unref` here to get rid of it.
net_buf_unref(target_buf);

bt_send_whatever(view);
// But we don't do `net_buf_unref(view)`, since here the ref was moved.
// That's inconsistent!
```

Note: There already is `net_buf_unref(view->parent)` in `destroy_view()`.

Reply:
The extra `net_buf_unref` can also cause a function that would have been applicable for tail-call optimization to not be. (In principle - I have not tested this, and it depends on ABI details, etc.)

Reply:
Moving the ownership of the reference is fine with me. However when doing that, I think it's good practice to set the original pointer to NULL to make the intent clear (that you're not using it anymore). The reference count is, as the name implies, supposed to count references, i.e. the number of pointer variables pointing at the same object. Setting the original pointer variable to NULL makes sure that this definition of what a reference count is still holds true.

Reply:
This reminds me, I think there was at some point discussion of having an explicit API to move references that takes a pointer to a pointer and sets the original value to NULL.

Reply:
inline function might be best (giving you type checks which macros don't do). There's at least one place already that could use it:
https://github.com/zephyrproject-rtos/zephyr/blob/59212c198cf60f5c1380bb4a21eb9ff059ed4f85/subsys/bluetooth/host/hci_core.c#L2348
If we want to take this even further, we could consider updating all our APIs which move the reference to take a pointer to pointer and clear the original value.

Reply:
made this https://github.com/zephyrproject-rtos/zephyr/issues/72798